geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java

```diff
@@ -237,15 +238,21 @@ public int hstrlen(byte[] field) {
     return new ArrayList<>(hash.keySet());
   }
 
-  public ImmutablePair<Integer, List<byte[]>> hscan(Pattern matchPattern,
-      int count,
-      int cursor) {
-
-    ArrayList<byte[]> resultList = new ArrayList<>(count + 2);
+  public ImmutablePair<Integer, List<byte[]>> hscan(Pattern matchPattern, int count, int cursor) {
+    // No need to allocate more space than it's possible to use given the size of the hash. We need
+    // to add 1 to hash.size() to ensure that if count > hash.size(), we return a cursor of 0
+    long maximumCapacity = 2L * Math.min(count, hlen() + 1);
+    if (maximumCapacity > Integer.MAX_VALUE) {
+      LogService.getLogger().error(
+          "The size of the data to be returned by hscan, {}, exceeds the maximum capacity of an array",
+          maximumCapacity);
+      throw new OutOfMemoryError("Requested array size exceeds VM limit");
```

Comment:
I think OutOfMemoryError should be reserved for when you actually run out of heap. In this case you are preventing that from happening by not doing a large allocation. Some code will no longer trust the state of the jvm if it sees an OutOfMemoryError and will initiate a shutdown. You could use IllegalArgumentException instead. Whatever exception is thrown here should it be caught in HScanExecutor and converted to RedisResponse.error?

Reply:
Currently, the exception thrown here will ultimately get handled in `ExecutionHandlerContext.getExceptionResponse()` where it will result in an error message returned to the client with the message "The server had an internal error please try again". This definitely doesn't seem correct, since trying again will just result in the same error, but none of the currently existing error responses we have seem right either, and I'm having a hard time find a list of existing Redis error messages to see if any of them would be a better fit. It's also not practical to test what Redis actually does in this situation, given that would require a Redis hash with over a billion entries. I'll have a look at the Redis source and see if I can figure out what the behaviour would be in this situation.

Reply:
How about ERROR_NOT_INTEGER? In this case the size is "out of range". Also be sure not to use OutOfMemoryError.

Reply:
Sounds good to me. I checked Redis' code and they don't hit this issue, since they're able to allocate lists with a maximum length of Long.MAX_VALUE, which is greater than their maximum number of entries in a hash, so this is one case where we're not able to exactly match their behaviour. Thankfully, it's also a case that should never realistically come up, since it's highly unlikely that users will be putting over a billion entries into one Redis hash.